File: src/lib/strengthProgress.ts
import type { AppData, JointCheckin, WorkoutLog } from './types'

export type JointRegion = 'arms' | 'core' | 'legs'
export type DeloadState = 'clear' | 'watch' | 'regional_deload' | 'whole_deload' | 'stop_and_review'

export interface StrengthPoint {
  sessionId: string
  date: string
  topWeight: number
  estimated1rm: number
  volume: number
  setWeights: Record<number, number>
}

export interface ExerciseStrengthSeries {
  key: string
  exerciseId: string | null
  name: string
  points: StrengthPoint[]
}

export interface SessionStrengthInsight {
  key: string
  name: string
  current: StrengthPoint
  previous: StrengthPoint | null
  reference: StrengthPoint | null
  daysCompared: number | null
  loadDelta: number | null
  estimated1rmDelta: number | null
}

export interface JointAssessment {
  state: DeloadState
  affected: JointRegion[]
  average: number
  highest: number
  rising: JointRegion[]
}

function exerciseKey(log: Pick<WorkoutLog, 'exercise_id' | 'exercise_name'>): string {
  return log.exercise_id ? `id:${log.exercise_id}` : `name:${log.exercise_name.trim().toLocaleLowerCase('en')}`
}

function daysBetween(earlier: string, later: string): number {
  return Math.max(0, Math.round((Date.parse(`${later}T12:00:00Z`) - Date.parse(`${earlier}T12:00:00Z`)) / 86_400_000))
}

export function estimatedOneRepMax(weight: number, reps: number | null): number {
  if (!Number.isFinite(weight) || weight <= 0) return 0
  if (reps == null || reps <= 1) return Math.round(weight * 10) / 10
  /* Epley is intentionally capped at 15 reps; higher-rep sets are poor
     maximal-strength estimators and should not create spectacular fake gains. */
  const boundedReps = Math.min(15, reps)
  return Math.round(weight * (1 + boundedReps / 30) * 10) / 10
}

export function buildStrengthSeries(data: AppData): ExerciseStrengthSeries[] {
  const sessions = new Map(data.workout_sessions.filter((session) => session.completed).map((session) => [session.id, session]))
  const grouped = new Map<string, {
    exerciseId: string | null
    name: string
    sessions: Map<string, WorkoutLog[]>
  }>()

  for (const log of data.workout_logs) {
    if (log.skipped || log.weight_kg == null || log.weight_kg <= 0 || !sessions.has(log.session_id)) continue
    const key = exerciseKey(log)
    const group = grouped.get(key) ?? {
      exerciseId: log.exercise_id,
      name: log.exercise_name,
      sessions: new Map<string, WorkoutLog[]>(),
    }
    const logs = group.sessions.get(log.session_id) ?? []
    logs.push(log)
    group.sessions.set(log.session_id, logs)
    grouped.set(key, group)
  }

  return [...grouped.entries()]
    .map(([key, group]) => {
      const points = [...group.sessions.entries()].map(([sessionId, logs]) => {
        const session = sessions.get(sessionId)!
        const usable = logs.filter((log) => log.weight_kg != null && log.weight_kg > 0)
        const topWeight = Math.max(...usable.map((log) => Number(log.weight_kg)))
        const estimated1rm = Math.max(...usable.map((log) => estimatedOneRepMax(Number(log.weight_kg), log.reps)))
        const volume = usable.reduce((sum, log) => sum + Number(log.weight_kg) * Math.max(0, log.reps ?? 0), 0)
        const setWeights = Object.fromEntries(usable.map((log) => [log.set_no, Number(log.weight_kg)]))
        return { sessionId, date: session.date, topWeight, estimated1rm, volume, setWeights }
      }).sort((a, b) => a.date.localeCompare(b.date) || a.sessionId.localeCompare(b.sessionId))
      return { key, exerciseId: group.exerciseId, name: group.name, points }
    })
    .filter((series) => series.points.length > 0)
    .sort((a, b) => a.name.localeCompare(b.name))
}

export function sessionStrengthInsights(data: AppData, sessionId: string): SessionStrengthInsight[] {
  const currentSession = data.workout_sessions.find((session) => session.id === sessionId)
  if (!currentSession) return []

  return buildStrengthSeries(data).flatMap((series) => {
    const current = series.points.find((point) => point.sessionId === sessionId)
    if (!current) return []
    const prior = series.points.filter((point) => point.date < currentSession.date)
    const previous = prior.at(-1) ?? null
    /* Prefer a meaningful 30-90 day comparison. If history is newer, the
       earliest prior point is still more honest than fabricating a horizon. */
    const inNinetyDays = prior.filter((point) => daysBetween(point.date, currentSession.date) <= 90)
    const reference = inNinetyDays[0] ?? previous
    return [{
      key: series.key,
      name: series.name,
      current,
      previous,
      reference,
      daysCompared: reference ? daysBetween(reference.date, currentSession.date) : null,
      loadDelta: reference ? Math.round((current.topWeight - reference.topWeight) * 10) / 10 : null,
      estimated1rmDelta: reference ? Math.round((current.estimated1rm - reference.estimated1rm) * 10) / 10 : null,
    }]
  }).sort((a, b) => (b.loadDelta ?? -Infinity) - (a.loadDelta ?? -Infinity))
}

export function assessJointCheckin(current: JointCheckin, previous?: JointCheckin | null): JointAssessment {
  const regions: JointRegion[] = ['arms', 'core', 'legs']
  const scores = regions.map((region) => current[region])
  const affected = regions.filter((region) => current[region] >= 5)
  const rising = previous
    ? regions.filter((region) => current[region] - previous[region] >= 2)
    : []
  const highest = Math.max(...scores)
  const average = Math.round((scores.reduce((sum, score) => sum + score, 0) / scores.length) * 10) / 10
  const severe = regions.filter((region) => current[region] >= 9)
  const high = regions.filter((region) => current[region] >= 7)
  const elevated = regions.filter((region) => current[region] >= 6)

  let state: DeloadState = 'clear'
  if (severe.length > 0) state = 'stop_and_review'
  else if (high.length >= 2 || elevated.length >= 2 || average >= 7) state = 'whole_deload'
  else if (high.length === 1) state = 'regional_deload'
  else if (affected.length > 0 || rising.length > 0) state = 'watch'

  return { state, affected: severe.length > 0 ? severe : affected, average, highest, rising }
}

export function checkinDue(checkins: JointCheckin[], today: string, intervalDays = 7): boolean {
  const latest = [...checkins].sort((a, b) => b.date.localeCompare(a.date))[0]
  return !latest || daysBetween(latest.date, today) >= intervalDays
}

File: ios/APEXNative/APEX/Core/Engine/StrengthProgress.swift
import Foundation

/*
 * Port of src/lib/strengthProgress.ts.
 *
 * Turns completed sets into a strength history: an estimated one-rep max per
 * exercise per session, what changed since last time, and what the weekly
 * joint check-in says about whether to keep pushing.
 */
enum StrengthProgress {
    enum JointRegion: String, Sendable, CaseIterable {
        case arms, core, legs
    }

    enum DeloadState: String, Sendable {
        case clear
        case watch
        case regionalDeload = "regional_deload"
        case wholeDeload = "whole_deload"
        case stopAndReview = "stop_and_review"
    }

    struct Point: Hashable, Sendable {
        let sessionID: UUID
        let date: String
        let topWeight: Double
        let estimated1RM: Double
        let volume: Double
        let setWeights: [Int: Double]
    }

    struct Series: Hashable, Sendable {
        let key: String
        let exerciseID: UUID?
        let name: String
        let points: [Point]
    }

    struct SessionInsight: Hashable, Sendable {
        let key: String
        let name: String
        let current: Point
        let previous: Point?
        /// The point the comparison is made against, preferring 30-90 days back.
        let reference: Point?
        let daysCompared: Int?
        let loadDelta: Double?
        let estimated1RMDelta: Double?
    }

    struct JointCheckin: Hashable, Sendable {
        let date: String
        let arms: Int
        let core: Int
        let legs: Int

        subscript(region: JointRegion) -> Int {
            switch region {
            case .arms: arms
            case .core: core
            case .legs: legs
            }
        }
    }

    struct JointAssessment: Hashable, Sendable {
        let state: DeloadState
        let affected: [JointRegion]
        let average: Double
        let highest: Int
        let rising: [JointRegion]
    }

    // MARK: - Helpers

    /// JavaScript's Math.round rounds half up; Swift rounds half away from
    /// zero. Deltas here can be negative, so the JS rule is kept exactly.
    private static func roundJS(_ value: Double) -> Double {
        (value + 0.5).rounded(.down)
    }

    private static func round1(_ value: Double) -> Double {
        roundJS(value * 10) / 10
    }

    private static func exerciseKey(_ log: WorkoutLog) -> String {
        if let id = log.exerciseID { return "id:\(id.uuidString.lowercased())" }
        return "name:\(log.exerciseName.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
    }

    private static func daysBetween(_ earlier: String, _ later: String) -> Int {
        guard let start = APEXDateMath.date(from: earlier), let end = APEXDateMath.date(from: later) else { return 0 }
        return max(0, Int(roundJS(end.timeIntervalSince(start) / 86_400)))
    }

    // MARK: - Estimated one-rep max

    static func estimatedOneRepMax(weight: Double, reps: Int?) -> Double {
        guard weight.isFinite, weight > 0 else { return 0 }
        guard let reps, reps > 1 else { return roundJS(weight * 10) / 10 }
        /* Epley is capped at 15 reps on purpose: higher-rep sets estimate a
           maximum poorly and would manufacture spectacular fake gains. */
        let bounded = Double(min(15, reps))
        return roundJS(weight * (1 + bounded / 30) * 10) / 10
    }

    // MARK: - Series

    static func buildSeries(sessions allSessions: [WorkoutSession], logs: [WorkoutLog]) -> [Series] {
        var sessions: [UUID: WorkoutSession] = [:]
        for session in allSessions where session.completed { sessions[session.id] = session }

        struct Group {
            var exerciseID: UUID?
            var name: String
            var sessions: [UUID: [WorkoutLog]] = [:]
            var order: [UUID] = []
        }
        var grouped: [String: Group] = [:]
        var groupOrder: [String] = []

        for log in logs {
            guard !log.skipped, let weight = log.weightKG, weight > 0, sessions[log.sessionID] != nil else { continue }
            let key = exerciseKey(log)
            if grouped[key] == nil {
                grouped[key] = Group(exerciseID: log.exerciseID, name: log.exerciseName)
                groupOrder.append(key)
            }
            if grouped[key]?.sessions[log.sessionID] == nil { grouped[key]?.order.append(log.sessionID) }
            grouped[key]?.sessions[log.sessionID, default: []].append(log)
        }

        return groupOrder.compactMap { key -> Series? in
            guard let group = grouped[key] else { return nil }
            let points = group.order.compactMap { sessionID -> Point? in
                guard let session = sessions[sessionID], let logs = group.sessions[sessionID] else { return nil }
                let usable = logs.filter { ($0.weightKG ?? 0) > 0 }
                guard !usable.isEmpty else { return nil }
                let topWeight = usable.compactMap(\.weightKG).max() ?? 0
                let best = usable.map { estimatedOneRepMax(weight: $0.weightKG ?? 0, reps: $0.reps) }.max() ?? 0
                let volume = usable.reduce(0.0) { $0 + ($1.weightKG ?? 0) * Double(max(0, $1.reps ?? 0)) }
                var setWeights: [Int: Double] = [:]
                for log in usable { setWeights[log.setNumber] = log.weightKG ?? 0 }
                return Point(
                    sessionID: sessionID, date: session.date, topWeight: topWeight,
                    estimated1RM: best, volume: volume, setWeights: setWeights
                )
            }.sorted {
                $0.date == $1.date
                    ? $0.sessionID.uuidString.lowercased() < $1.sessionID.uuidString.lowercased()
                    : $0.date < $1.date
            }
            guard !points.isEmpty else { return nil }
            return Series(key: key, exerciseID: group.exerciseID, name: group.name, points: points)
        }.sorted { $0.name < $1.name }
    }

    // MARK: - Session insights

    static func sessionInsights(
        sessions: [WorkoutSession],
        logs: [WorkoutLog],
        sessionID: UUID
    ) -> [SessionInsight] {
        guard let current = sessions.first(where: { $0.id == sessionID }) else { return [] }

        return buildSeries(sessions: sessions, logs: logs).compactMap { series -> SessionInsight? in
            guard let point = series.points.first(where: { $0.sessionID == sessionID }) else { return nil }
            let prior = series.points.filter { $0.date < current.date }
            let previous = prior.last
            /* Prefer a meaningful 30-90 day comparison. When history is newer,
               the earliest prior point is still more honest than inventing a
               horizon that does not exist. */
            let withinNinety = prior.filter { daysBetween($0.date, current.date) <= 90 }
            let reference = withinNinety.first ?? previous
            return SessionInsight(
                key: series.key,
                name: series.name,
                current: point,
                previous: previous,
                reference: reference,
                daysCompared: reference.map { daysBetween($0.date, current.date) },
                loadDelta: reference.map { round1(point.topWeight - $0.topWeight) },
                estimated1RMDelta: reference.map { round1(point.estimated1RM - $0.estimated1RM) }
            )
        }.sorted { ($0.loadDelta ?? -.infinity) > ($1.loadDelta ?? -.infinity) }
    }

    // MARK: - Joint check-ins

    static func assess(_ current: JointCheckin, previous: JointCheckin? = nil) -> JointAssessment {
        let regions = JointRegion.allCases
        let scores = regions.map { current[$0] }
        let affected = regions.filter { current[$0] >= 5 }
        let rising = previous.map { prior in regions.filter { current[$0] - prior[$0] >= 2 } } ?? []
        let highest = scores.max() ?? 0
        let average = roundJS(Double(scores.reduce(0, +)) / Double(scores.count) * 10) / 10
        let severe = regions.filter { current[$0] >= 9 }
        let high = regions.filter { current[$0] >= 7 }
        let elevated = regions.filter { current[$0] >= 6 }

        var state: DeloadState = .clear
        if !severe.isEmpty { state = .stopAndReview }
        else if high.count >= 2 || elevated.count >= 2 || average >= 7 { state = .wholeDeload }
        else if high.count == 1 { state = .regionalDeload }
        else if !affected.isEmpty || !rising.isEmpty { state = .watch }

        return JointAssessment(
            state: state,
            affected: severe.isEmpty ? affected : severe,
            average: average,
            highest: highest,
            rising: rising
        )
    }

    static func checkinDue(_ checkins: [JointCheckin], today: String, intervalDays: Int = 7) -> Bool {
        guard let latest = checkins.max(by: { $0.date < $1.date }) else { return true }
        return daysBetween(latest.date, today) >= intervalDays
    }

    /// Check-ins live in settings addons, the same place the web keeps them.
    static func checkins(from addons: [String: JSONValue]?) -> [JointCheckin] {
        guard case .array(let rows)? = addons?["joint_checkins"] else { return [] }
        return rows.compactMap { row in
            guard case .object(let fields) = row,
                  case .string(let date)? = fields["date"] else { return nil }
            func score(_ key: String) -> Int {
                guard case .number(let value)? = fields[key] else { return 0 }
                return max(0, min(10, Int(value)))
            }
            return JointCheckin(date: date, arms: score("arms"), core: score("core"), legs: score("legs"))
        }
    }
}
